src/coreclr/vm/threadsuspend.cpp

```diff
@@ -2896,9 +2896,15 @@ BOOL Thread::RedirectThreadAtHandledJITCase(PFN_REDIRECTTARGET pTgt)
     if (!pCtx)
     {
         pCtx = m_pSavedRedirectContext = ThreadStore::GrabOSContext(&m_pOSContextBuffer);
-        _ASSERTE(GetSavedRedirectContext() != NULL);
     }
 
+    // We may not have a preallocated context. Could be short on memory when we tried to preallocate.
```

Comment:
The crash was hit in the CI by a lot of tests. It sounds very unlikely we would be running out of memory that often.

Isn't the actual problem a race condition somewhere?

Reply:
I was unable to reproduce this so far. Did we see this on anything other than Windows7/x86?

Reply:
I think we have seen it on Win7/x86 only

Reply:
We call `InitializeContext` with 0 for context buffer size to get the actual buffer size. The expected result is to get `ERROR_INSUFFICIENT_BUFFER and to get the size required. 
Rarely, for unknown reasons we get a different error. We have disabled an assert because of this.

https://github.com/dotnet/runtime/blob/1857d04600c098475db8dc13bf98007b68761033/src/coreclr/vm/threadsuspend.cpp#L1986

I wonder if in such case we also get invaid size (like 0) and thus the allocation fails.

Reply:
The spec apparently mentions that:

```
 If the function fails with an error other than ERROR_INSUFFICIENT_BUFFER, the contents of ContextLength are undefined.
```

Reply:
I just do not know what can possibly go wrong with a call like `InitializeContext(NULL, context, NULL, &contextSize);`   - that is the same use as in the example.

Reply:
I think we should not allocate when `InitializeContext` returns unexpected results.

Reply:
Perhaps that is the root cause, but the reason why the call may fail on Win7/x86 is still unclear.